src/main/scala/is/hail/annotations/UnsafeAnnotations.scala

```diff
@@ -0,0 +1,402 @@
+package is.hail.annotations
+
+import is.hail.expr._
+import is.hail.utils._
+import is.hail.variant.{AltAllele, Variant}
+import org.apache.spark.sql.Row
+import org.apache.spark.unsafe.Platform
+
+import scala.collection.mutable.ArrayBuffer
+
+object UnsafeAnnotations {
+  def missingBytes(elems: Int): Int = (elems + 31) / 32 * 4
+
+  def arrayElementSize(t: Type): Int = {
+    var eltSize = t.byteSize
+    val mod = eltSize % t.alignment
+    if (mod != 0)
+      eltSize += (t.alignment - mod)
+    eltSize
+  }
+
+  def roundUpAlignment(offset: Int): Int = {
+    val mod = offset & 0x7
+    if (mod != 0)
+      offset + (8 - mod)
+    else
+      offset
+  }
+}
+
+class UnsafeRowBuilder(t: TStruct, sizeHint: Int = 0, debug: Boolean = false) {
```

Comment:
Do you mean temporally or spatially?  I guess the answer is yes to both, first temporally, we should reuse MemoryBuffers for successive records, this is the same as the mutable stuff we started on, and then spatially we can decode and process multiple records at a time instaed of using iterators per-record (what database people call vectorization).